APA102.class.nut

```diff
@@ -0,0 +1,169 @@
+    // plus head and tail frames for APA102 signalling. The head frame is
+    // four bytes of 0x00, the tail is four bytes of 0xFF (see constructor)
+    _data = null;
+
+    function constructor(spiBus = null, numPixels = 0, clockPin = null, dataPin = null) {
+        // To select your mode, pass in a valid imp SPI bus OR valid imp GPIO pins (clock and data)
+        // SPI mode is the default (_mode is true)
+        if (numPixels == 0) {
+            server.error("APA102() requires a non-zero number of pixels.");
+            return null;
+        } else {
+            _num = numPixels;
+        }
+
+        if (spiBus == null) {
+            // Selecting GPIO 'bit bang' mode
+            if (clockPin == null || dataPin == null) {
+                server.error("APA102() requires either a non-null SPI bus, or clock and data pin objects.");
+                return null;
+            } else {
+                _clk = clockPin;
+                _clk.configure(DIGITAL_OUT, 0);
+                _dat = dataPin;
+                _dat.configure(DIGITAL_OUT, 0);
+                _mode = false;
+            }
+        } else {
+            _spi = spiBus;
+        }
+
+        // Set up the data frame store
+        _data = blob(4 + _num * FRAME_LENGTH + 4);
+
+        // Create sentinel frames: the first frame (four bytes) is all 0s, the final one is all 1s
+        _data.seek(0, 'b');
+        _data.writen(0x00, 'i');
+        _data.seek(4 + numPixels * FRAME_LENGTH, 'b');
+        _data.writen(0xFFFFFFFF, 'i');
+
+        // Turn off all the LEDs
+        fill([0,0,0]).draw();
+    }
+
+    function configure() {
+        // This is optional but for SPI mode only
+        // Note that the imp will select the highest speed at 18000kHz or under
+        _spi.configure(SIMPLEX_TX, 24000);
+        return this;
+    }
+
+    function set(pixel, color) {
+        // color is an array: [r, g, b] or [r, g, b, brightness]
+        // Default to full brightness if not specified
+
+        if (pixel < 0 || pixel > _num - 1) {
+            server.error("APA102.set() selected pixel is out of range");
+            return this;
+        }
+
+        local brightness = color.len() == 4 ? color[3] : 0xFF;
+        _data.seek(4 + pixel * FRAME_LENGTH, 'b');
+        _data.writeblob(_makeFrame(brightness, color[0], color[1], color[2]));
+        return this;
+    }
+
+    function fill(color, start = 0, end = null) {
+        if (start < 0 || start > _num - 1) {
+            server.error("APA102.fill() start parameter is out of range");
+            return this;
+        }
+
+        if (end != null && end < start) {
+            server.error("APA102.fill() end parameter is out of range");
+            return this;
+        }
+
+        local endIndex = (end == null) ? _num - 1 : end;
+        if (endIndex >= _num) endIndex = _num - 1;
+
+        local brightness = color.len() == 4 ? color[3] : 0xFF;
+
+        for (local i = start ; i <= endIndex ; ++i) {
```

Comment:
I usually do ++i because (I read) it uses fewer memory accesses, ie. updates variable in memory directly rather that creates a new variable in which the incremented value is placed. But fair point, that may not be the case with Squirrel.